presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java

```diff
@@ -75,6 +64,29 @@ public static Block bigintUnion(@SqlType("array(bigint)") Block leftArray, @SqlT
         return distinctElementBlockBuilder.build();
     }
 
+    @SqlType("array(boolean)")
+    public static Block booleanUnion(@SqlType("array(boolean)") Block leftArray, @SqlType("array(boolean)") Block rightArray)
```

Comment:
I'd write this whole thing as follows. No need to additional classes or containers.

```java
boolean hasNull = false;
boolean hasTrue = false;
boolean hasFalse = false;

for (int i = 0; i < left.getPositionCount(); i++) {
    if (left.isNull(i)) {
        hasNull = true;
    }
    else if (BOOLEAN.getBoolean(left, i)) {
        hasTrue = true;
    }
    else {
        hasFalse = true;
    }
}

for (int i = 0; i < right.getPositionCount(); i++) {
    if (right.isNull(i)) {
        hasNull = true;
    }
    else if (BOOLEAN.getBoolean(right, i)) {
        hasTrue = true;
    }
    else {
        hasFalse = true;
    }
}

if (hasNull) {
    distinctElementBlockBuilder.appendNull();
}
if (hasTrue) {
    BOOLEAN.writeBoolean(distinctElementBlockBuilder, true);
}
if (hasFalse) {
    BOOLEAN.writeBoolean(distinctElementBlockBuilder, false);
}

return distinctElementBlockBuilder.build();
```

Reply:
@martint This was my originial writing except the part I removed the branching for true and false. Then I thought people don't like repeated code so come up with current version. I can change it back.

Reply:
Repeated code that's reasonably simple is not a problem. It has to be considered in the context of how much additional complexity and "conceptual weight" the alternatives introduce (new classes, interfaces, APIs, indirections, code that cannot be understood solely by looking at a localized area, etc)

Reply:
If you want to avoid repeating the loops, you can always do this like this. But I think the unrolled version (@martint 's) is better.
```
for (Block block : ImmutableList.of(left, right)) {
   for (int i = 0; i < block.getPositionCount(); i++) {
       if (left.isNull(i)) {
           hasNull = true;
       ....
}
```

@martint in this context, would it make sense to break out of the `for` loop as soon as `hasNull`, `hasTrue` and `hasFalse` are all set?

Reply:
Very possibly, but it will depend on whether the extra checks are cheap on average compared to plowing through all the items. Some benchmarking would give us the answer.

Reply:
And on the data -- for long arrays with mixed values it would be huge speedup. For long arrays without null, false or true, it will, by necessity, be slower. Benchmark would tell how much.